Provide CipherSuites option between Clients and Servers

lzhecheng opened the pull request:
* Add options TLSCipherSuites, TLSMinVersion on Apiserver side of controller
  and agent.
* Add a util module cipher to support functionality and related UT.
* Add tls e2e tests to verify Apiserver of Antrea and Antrea agent.

Signed-off-by: Zhecheng <lzhecheng@vmware.com>

jianjuns commented on cmd/antrea-agent/config.go:
```diff
@@ -127,4 +127,10 @@ type AgentConfig struct {
 	// Provide the address of Kubernetes apiserver, to override any value provided in kubeconfig or InClusterConfig.
 	// Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
 	KubeAPIServerOverride string `yaml:"kubeAPIServerOverride,omitempty"`
+	// Cipher suites to use on Client side.
```
Do you use the upper case for "Client" on purpose?

lzhecheng replied:
Yes.

jianjuns commented on pkg/util/cipher/cipher.go:
```diff
@@ -0,0 +1,64 @@
+	"VersionTLS13": tls.VersionTLS13,
+}
+
+// NewTLSConfig generates a *tls.Config with cipher suites string, TLS min version and TLS max version.
```
Probably remove "*" of "*tls.Config".

lzhecheng replied:
Updated.

jianjuns reviewed (commented):
LGTM.

antoninbas commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -120,3 +120,12 @@ featureGates:
 # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
 #kubeAPIServerOverride: ""
+
+# Comma-separated list of cipher suites.
```
Maybe we could match the K8s documentation here and add "If omitted, the default Go cipher suites will be used." And even add this link: https://golang.org/pkg/crypto/tls/#CipherSuites

lzhecheng replied:
Updated. But I prefer this link since this one includes possible cipher suite value: https://golang.org/pkg/crypto/tls/#pkg-constants

antoninbas commented on pkg/util/cipher/cipher.go:
```diff
@@ -0,0 +1,64 @@
+	"VersionTLS13": tls.VersionTLS13,
+}
+
+// NewTLSConfig generates a *tls.Config with cipher suites string, TLS min version and TLS max version.
```
s/with/given

lzhecheng replied:
Updated to `tls.Config with given cipher suites ...`
Not sure if this is your intention or `tls.Config given cipher suite ...` but it sounds better for me.

antoninbas commented on pkg/util/cipher/cipher.go:
```diff
@@ -0,0 +1,64 @@
+	}, nil
+}
+
+// AppendTLSConfig appends TLS config (Cipher Suites, caBundle, serverName) into *rest.Config.
```
s/appends into/appends to

lzhecheng replied:
Updated.

antoninbas reviewed (commented):
some minor comments, otherwise LGTM

I am just curious about why we are adding these parameters on the Agent / client side, and not adding them to the Controller to configure the apiserver?

lzhecheng commented:
@antoninbas Yes, I only configured client side. The reason is that I added `PreferServerCipherSuites: false` so it should work. I also verified it with the e2e test I added.
Please share you concern if any.

antoninbas commented:
> @antoninbas Yes, I only configured client side. The reason is that I added `PreferServerCipherSuites: false` so it should work. I also verified it with the e2e test I added.
> Please share you concern if any.

I think I was surprised because I thought we would add these flags for the servers, and not the clients. Similar to how these flags are available for kube-apiserver.

But my concern is also for APIServices. Don't you want a way to configure the TLS config for communications between kube-apiserver and the Antrea Controller apiserver? Or is this something you configure in K8s already?
Same thing for communications between antctl and the Agent / Controller apiserver.

lzhecheng commented:
24

> > @antoninbas Yes, I only configured client side. The reason is that I added `PreferServerCipherSuites: false` so it should work. I also verified it with the e2e test I added.
> > Please share you concern if any.
> 
> I think I was surprised because I thought we would add these flags for the servers, and not the clients. Similar to how these flags are available for kube-apiserver.
> 
> But my concern is also for APIServices. Don't you want a way to configure the TLS config for communications between kube-apiserver and the Antrea Controller apiserver? Or is this something you configure in K8s already?
> Same thing for communications between antctl and the Agent / Controller apiserver.

In my mind, if someone wants to configure cipher suite, both k8s setup and this cipher suite option in Antrea will be configured. Then communication: "k8s apiserver - antrea apiserver" and "antrea apiserver - antrea client" will be ok. 

As for antctl, I did try to add cipher suite option but it turned out that "insecure: true" cannot be set at the same time. I think maybe it is not needed by design.
https://github.com/vmware-tanzu/antrea/blob/main/pkg/antctl/client.go#L71

I am not so familiar with TLS cipher suite. So discussion on the design is highly welcomed. :)

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,51 @@
+package e2e
```
license header

lzhecheng replied:
Updated.

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,51 @@
+	respCS := resp.TLS.CipherSuite
+	defer resp.Body.Close()
+
+	assert.Equal(t, cs, respCS, "Cipher Suite used by Server should be %s", tls.CipherSuiteName(cs))
```
Not sure if the test really cover any antrea logic. The PR doesn't change anything on server side but it's testing server response. Should it test agent can connect to controller when CipherSuites/ MinVersion/ MaxVersion is set?

lzhecheng replied:
I use this test to mock the behavior on client side, with ciphersuite, tlsversion, PreferServerCipherSuites.

tnqn commented on cmd/antrea-agent/agent.go:
```diff
@@ -71,9 +72,13 @@ func run(o *Options) error {
 
 	// Create Antrea Clientset for the given config.
-	antreaClientProvider := agent.NewAntreaClientProvider(o.config.AntreaClientConnection, k8sClient)
+	antreaClientProvider := agent.NewAntreaClientProvider(o.config.AntreaClientConnection, k8sClient, tlsConfig)
```
Don't it need to set same TLS config for agent-apiserver communication?

tnqn commented:
> > > @antoninbas Yes, I only configured client side. The reason is that I added `PreferServerCipherSuites: false` so it should work. I also verified it with the e2e test I added.
> > > Please share you concern if any.
> > 
> > 
> > I think I was surprised because I thought we would add these flags for the servers, and not the clients. Similar to how these flags are available for kube-apiserver.
> > But my concern is also for APIServices. Don't you want a way to configure the TLS config for communications between kube-apiserver and the Antrea Controller apiserver? Or is this something you configure in K8s already?
> > Same thing for communications between antctl and the Agent / Controller apiserver.
> 
> In my mind, if someone wants to configure cipher suite, both k8s setup and this cipher suite option in Antrea will be configured. Then communication: "k8s apiserver - antrea apiserver" and "antrea apiserver - antrea client" will be ok.
> 
> As for antctl, I did try to add cipher suite option but it turned out that "insecure: true" cannot be set at the same time. I think maybe it is not needed by design.
> https://github.com/vmware-tanzu/antrea/blob/main/pkg/antctl/client.go#L71
> 
> I am not so familiar with TLS cipher suite. So discussion on the design is highly welcomed. :)

I also have same question as @antoninbas. Even `PreferServerCipherSuites: false` can make them communicate with the cipher suite the client chooses. There are a few reasons that may make server side configuration prefered:
1. Configuring it on server side can prevent insecure cipher suites from being used by any clients. It solves the issue of antctl, kube-apiserver as clients.
2. From the code, it seems the client library is not friendly to set cipher suite, while the server library is very friendly for this.
3. Guess the current e2e will only make sense when implementing it on server side, otherwise it's just testing http library.

tnqn commented:
And for maxVersion, is it really useful? I see typically only minVersion is configurable.

lzhecheng commented on pkg/util/cipher/cipher.go:
```diff
@@ -0,0 +1,67 @@
+	if err != nil {
+		return nil, err
+	}
+	// #nosec G402: ignore MinVersion and MaxVersion options in test code
```
To ignore error for golangci-int. It says there's error on TLS MinVersion and MaxVersion.

lzhecheng commented on pkg/util/cipher/cipher.go:
```diff
@@ -0,0 +1,67 @@
+
+// AppendTLSConfig appends TLS config (Cipher Suites, TLSVersion, PreferServerCipherSuites, caBundle, serverName) to
+// *rest.Config.
+func AppendTLSConfig(c *rest.Config, cipherTLSConfig *tls.Config, caBundle []byte, serverName string) error {
```
Actually not, `c.Transport` is set at L65 with all those options (CipherSuites, TLSVersion).

lzhecheng commented:
> And for maxVersion, is it really useful? I see typically only minVersion is configurable.

I think usually TLS1.2 and TLS1.3 are used so minVersion can be configured to TLS1.2. But maybe someone prefers TLS1.2 then maxVersion can be set to TLS1.2.

lzhecheng commented:
For someone comes across this PR when configuring Cipher Suites for K8s, please pay attention that choose a proper one like "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256".

lzhecheng commented on cmd/antrea-agent/config.go:
```diff
@@ -127,4 +127,10 @@ type AgentConfig struct {
 	// Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
 	KubeAPIServerOverride string `yaml:"kubeAPIServerOverride,omitempty"`
+	// Cipher suites to use on Client side.
+	CipherSuites string `yaml:"cipherSuites,omitempty"`
```
Updated.

lzhecheng commented:
This PR is refactored to configure on apiserver side. PTAL.

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,159 @@
+	if len(svc.Spec.Ports) == 0 {
+		t.Fatal("Antrea Service has no ports")
+	}
+	url := fmt.Sprintf("https://%s:%d", svc.Spec.ClusterIP, svc.Spec.Ports[0].Port)
+
+	// 1. TLSMaxVersion unset, then a TLS1.3 Cipher Suite should be used.
+	respCipherSuite := request(t, url, 0)
```
you can't assume that the ClusterIP is accessible from outside the cluster, it's why the Kind tests are failing

you can use `openssl` like you did for the Agent apiserver below. Or maybe you can get the Pod IP for the antrea-agent / antrea-controller Pod, which would be the Node IP, since these are hostNetwork Pods. It's safe to assume that the Node IP can be reached from the test program, since we support running commands on Nodes over SSH. Either way, maybe we should unify the "request" method for the Controller apiserver and the Agent apiserver.

lzhecheng replied:
You're right. Code refactored.

antoninbas reviewed (commented):
Thanks for switching configuration from client to server

lzhecheng commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -120,3 +120,13 @@ featureGates:
+# Comma-separated list of cipher suites. If omitted, the default Go cipher suites will be used.
+# https://golang.org/pkg/crypto/tls/#pkg-constants
+# Note that Antrea is using old version of k8s.io/component-base, so TLS1.3 Cipher suites cannot be filled here but it
+# doesn't stop apiserver from using TLS1.3 Cipher Suites.
```
From my observation, TLS1.3 Cipher suites are always in apiserver's consideration and preferred, whether the list is empty or not. When `MaxTLSVersion` is explicitly set to TLS1.2, then TLS1.3 Cipher suites are not in consideration any more.
I am going to rephrase it to "so TLS1.3 Cipher suites cannot be explicitly listed here but the apiserver can still use TLS1.3 Cipher Suites in practice and as preferred".

antoninbas replied:
> TLS1.3 Cipher suites are always in apiserver's consideration and preferred, whether the list is empty or not

So if in the ConfigMap, I define `cipherSuites: <some TLS1.2 cipher suite>`, the apiserver will still prefer the TLS1.3 cipher suites over the one I explicitly provide?

lzhecheng replied:
Yes from my observation.

antoninbas replied:
Thanks for clarifying. It seems that this is actually the "desired" behavior for the Go crypto package (https://github.com/golang/go/issues/29349), and not something specific to K8s / kube-apiserver.

Given that, I would recommend some rephrasing:
What do you think of:
> Note that Antrea is using an old version of k8s.io/component-base, so TLS1.3 cipher suites cannot be explicitly listed here. Additionally, by default Antrea uses the Go "crypto" libraries, which do not support configuring cipher suites for TLS1.3. In other words, the apiserver will always prefer TLS1.3 cipher suites whenever possible.

lzhecheng replied:
Good idea. Thanks!

lzhecheng commented on pkg/util/cipher/cipher_test.go:
```diff
@@ -0,0 +1,44 @@
+		success bool
+	}{
+		{"TLS_RSA_WITH_RC4_128_SHA,TLS_RSA_WITH_3DES_EDE_CBC_SHA", []uint16{0x0005, 0x000a}, true},
+		{" TLS_RSA_WITH_RC4_128_SHA,   TLS_RSA_WITH_3DES_EDE_CBC_SHA ", []uint16{0x0005, 0x000a}, true},
```
Updated.

lzhecheng commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,159 @@
+		if len(cs.SupportedVersions) == 1 && cs.SupportedVersions[0] == tls.VersionTLS13 {
+			tls13CipherSuites[cs.ID] = true
+		}
+	}
```
It is removed now.

lzhecheng commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	masterNode := nodeName(0)
```
Updated. Thanks.

lzhecheng commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,149 @@
+
+func (data *TestData) configureTLS(t *testing.T, cipherSuites []uint16, tlsMinVersion string) {
+	var cipherSuitesStr string
+	if len(cipherSuites) > 0 {
```
nit: this is redundant with the loop.

lzhecheng replied:
Not sure what you mean. Are you suggesting `strings.Join()` here? `cipherSuites` is a uint16 list so `tls.CipherSuiteName()` is needed to turn it into string.

tnqn replied:
I mean the loop body won't be executed anyway without the check.

tnqn replied:
I mean the loop body won't be executed anyway without the check.

lzhecheng replied:
Got it. You're right. Updated.

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,149 @@
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#tlsMinVersion:", fmt.Sprintf("tlsMinVersion: %s", tlsMinVersion), 1)
+		data["antrea-agent.conf"] = antreaAgentConf
+	}, true, true); err != nil {
+		t.Fatalf("Failed to enable configure Cipher Suites and TLSMinVersion: %v", err)
```
```suggestion
		t.Fatalf("Failed to configure Cipher Suites and TLSMinVersion: %v", err)
```

lzhecheng replied:
Updated.

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,149 @@
+		{agentPodName, agentContainerName, apis.AntreaAgentAPIPort, "Agent"},
+	}
+	for _, tc := range tests {
+		data.checkTLS(t, tc.podName, tc.containerName, tc.apiserver, tc.apiserverStr)
```
It could create a subtest with the specified name to include context to the output of the test log. For example:
```
	tests := []struct {
                name          string
		podName       string
		containerName string
		apiserver     int
	}{
		{"Controller", controllerPodName, controllerContainerName, apis.AntreaControllerAPIPort},
		{"Agent", agentPodName, agentContainerName, apis.AntreaAgentAPIPort},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
		        data.checkTLS(t, tc.podName, tc.containerName, tc.apiserver)
```

lzhecheng replied:
Good idea. Updated.

tnqn commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,149 @@
+	stdouts = data.openssl(t, podName, containerName, true, apiserver)
+	for _, stdout := range stdouts {
+		assert.Equal(t, true, strings.Contains(stdout, fmt.Sprintf("New, TLSv1.2, Cipher is %s", cipherSuiteStr)),
+			"Cipher Suite used by %s Server should be the TLS1.2 one '%s', output: %s", apiserverStr, cipherSuiteStr, stdout)
```
Inconsistent with L107.

lzhecheng replied:
Updated.

tnqn reviewed (commented):
Overall LGTM, some nits

lzhecheng commented:
/test-e2e

lzhecheng commented:
/test-ipv6-only-e2e

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+		apiserver     int
+		apiserverStr  string
+	}{
+		{"TestControllerSideApiserver", controllerPodName, controllerContainerName, apis.AntreaControllerAPIPort, "Controller"},
+		{"TestAgentSideApiserver", agentPodName, agentContainerName, apis.AntreaAgentAPIPort, "Agent"},
```
maybe names of subtests should simply be "ControllerApiserver" and "AgentApiserver". The name of the parent test ("TestTLSCipherSuites") will be used as a prefix in log outputs

lzhecheng replied:
Updated.

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+	opensslTLS13CipherSuites = []string{"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256"}
+)
+
+// TestTLSCipherSuites tests Cipher Suite and TLSVersion config on Antrea Apiserver, Controller side or Agent side.
+func TestTLSCipherSuites(t *testing.T) {
```
what about "TestAntreaApiserverTLSConfig"?

lzhecheng replied:
Updated.

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+				break
+			}
+		}
+		assert.Equal(t, true, oneTLS13CS,
```
there is an `assert.True` assertion

lzhecheng replied:
Thank you. Updated.

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+	// 2. Set TLSMaxVersion to TLS1.2, then TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 should be used
+	stdouts = data.openssl(t, podName, containerName, true, apiserver)
+	for _, stdout := range stdouts {
+		assert.Equal(t, true, strings.Contains(stdout, fmt.Sprintf("New, TLSv1.2, Cipher is %s", cipherSuiteStr)),
```
same as above

lzhecheng replied:
Updated.

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+	opensslTLS13CipherSuites = []string{"TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256"}
+)
+
+// TestTLSCipherSuites tests Cipher Suite and TLSVersion config on Antrea Apiserver, Controller side or Agent side.
```
I don't think we usually capitalize "apiserver"

lzhecheng replied:
Used "apiserver"

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+
+func (data *TestData) openssl(t *testing.T, pod string, container string, tls12 bool, port int) []string {
+	var stdouts []string
+	tests := []struct {
```
suggest renaming `tests` to something else here, such as `opensslConnectCommands`

lzhecheng replied:
Updated.

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+		if tls12 {
+			cmd = append(cmd, "-tls1_2")
+		}
+		stdout, _, _ := data.runCommandFromPod(antreaNamespace, pod, container, cmd)
```
ignoring errors on purpose?

lzhecheng replied:
Yes, since we just need to the check Cipher Suite used.

antoninbas replied:
still it may be better to fail earlier if the command cannot be run for any reason (e.g. the Pod has crashed)

lzhecheng replied:
Got it. Better to be prepared for everthing. Updated.

antoninbas reviewed (commented):
some nits

antoninbas commented on test/e2e/tls_test.go:
```diff
@@ -0,0 +1,150 @@
+			"-6",
+		},
+	}
+	for _, tc := range opensslConnectCommands {
+		if !tc.enabled {
+			continue
+		}
+		cmd := []string{"timeout", "1", "openssl", "s_client", "-connect", net.JoinHostPort(tc.ip, fmt.Sprint(port)), tc.option}
+		if tls12 {
```
s/tc/c

lzhecheng replied:
Updated.

antoninbas reviewed (approved):
LGTM

jianjuns commented on build/yamls/antrea.yml:
```diff
@@ -1376,6 +1376,17 @@ data:
+
+    # Comma-separated list of cipher suites. If omitted, the default Go cipher suites will be used.
+    # https://golang.org/pkg/crypto/tls/#pkg-constants
+    # Note that Antrea is using an old version of k8s.io/component-base, so TLS1.3 cipher suites cannot be explicitly listed here.
```
I personally feel no need to explain the long reasons in yaml, but just describe the facts. We can move detailed explanation to comments on code. Up to you to decide.

antoninbas replied:
I feel that everything is needed, except for the `k8s.io/component-base` mention and the github link

jianjuns replied:
Yes, I meant "Antrea is using an old version of k8s.io/component-base" and "by default Antrea uses the Go "crypto" libraries".

antoninbas replied:
Sounds good. I had suggested using "by default Antrea uses the Go "crypto" libraries" because I was thinking of user which may have their own FIPS compliant builds of Antrea by using a different crypto implementation. But in that case, it's not really our responsibility to document the behavior.

lzhecheng replied:
Removed component-base and the link.

lzhecheng replied:
Also "Go crypto libraries"

jianjuns commented on build/yamls/antrea.yml:
```diff
@@ -1376,6 +1376,17 @@ data:
+    # https://golang.org/pkg/crypto/tls/#pkg-constants
+    # Note that Antrea is using an old version of k8s.io/component-base, so TLS1.3 cipher suites cannot be explicitly listed here.
+    # Additionally, by default Antrea uses the Go "crypto" libraries, which do not support configuring Cipher Suites for TLS1.3.
+    # In other words, the apiserver will always prefer TLS1.3 Cipher Suites whenever possible.
```
I do not understand this sentence - why "in other words", "the apiserver will always prefer TLS1.3 Cipher Suites whenever possible"? I cannot see the previous sentences are relevant to "the apiserver will always prefer TLS1.3 Cipher Suites whenever possible".

antoninbas replied:
https://github.com/golang/go/issues/29349

You cannot disable TLS1.3 Cipher Suites and they will always be preferred to the cipher suites you explicitly provide. So if you set `ltsCipherSuites` to `<some TLS1.2 suite>`, your choice will be ignored if the client supports TLS 1.3

jianjuns replied:
Should we just explicitly declare the two points: TLS1.3 cipher suites cannot be listed; but apiserver will always prefer TLS1.3 when it is supported by the client?

lzhecheng replied:
I think @jianjuns 's concern is that "in other words" isn't proper here. I changed it to "actually". Do you think it is better?

jianjuns commented on build/yamls/antrea-aks.yml:
```diff
@@ -1371,6 +1371,15 @@ data:
+
+    # Comma-separated list of Cipher Suites. If omitted, the default Go Cipher Suites will be used.
+    # https://golang.org/pkg/crypto/tls/#pkg-constants
+    # Note that TLS1.3 Cipher Suites cannot be explicitly listed here. Actually, the apiserver will
```
How about:

Note that TLS1.3 Cipher Suites cannot be added to the list. But the apiserver will always prefer TLS1.3 Cipher Suites whenever possible.

lzhecheng replied:
Updated.

lzhecheng commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all

lzhecheng commented:
/test-e2e

antoninbas reviewed (approved):
LGTM

lzhecheng commented:
ipv6-only-e2e is successful according to Jenkins output.
/test-ipv6-only-networkpolicy
/test-windows-conformance

lzhecheng commented:
/test-ipv6-only-networkpolicy

lzhecheng commented:
All tests are successful.
